hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java

```diff
@@ -0,0 +1,190 @@
+ */
+public class DelayerConnectionProvider implements
+		ConnectionProvider,
+		Configurable,
+		ServiceRegistryAwareService,
+		Stoppable {
+
+	private static final Logger LOGGER = Logger.getLogger( DelayerConnectionProvider.class );
+
+	private final ConnectionProvider connectionProvider;
+	private ServiceRegistry serviceRegistry;
+	// volatile due to the double checked lock
+	private volatile List<Runnable> operations = new ArrayList<>();
+	private final ReentrantLock lock = new ReentrantLock();
+
+	public DelayerConnectionProvider(ConnectionProvider connectionProvider) {
+		this.connectionProvider = connectionProvider;
+	}
+
+	/**
+	 * Operations can only be added before the {@code SessionFactory} is fully built.
+	 * This is to prevent concurrent additions of operations.
+	 * The operation must use the {@link ServiceRegistry} provided to retrieve
+	 * {@link ConnectionProvider} to get a connection.
+	 *
+	 * The operation will try and be executed eagerly, we might be lucky.
+	 *
+	 * @param operation on a connection ready database
+	 */
+	public void addOperation(Runnable operation) {
+		operations.add( operation );
+		tryAndExecuteOperations();
+	}
+
+	/**
+	 * Try to eagerly flush the operations if a connection is available
+	 */
+	private void tryAndExecuteOperations() {
+		// It would be more elegant to refactor the operation lock from getConnection is a shared method
+		// but I prefer the JVM to have more chances to keep getConnection() inlined
+		try {
+			Connection c = getConnection();
+			closeConnection( c );
+			if (operations == null) {
+				// operation execution has succeeded, but subsequent ones might happen
+				// remember, we are not multi-threaded here as we are bootstrapping Hibernate ORM
+				operations = new ArrayList<>();
+			}
+		}
+		catch (Exception e) {
+			// we don't care if there is a failure
+		}
+	}
+
+	@Override
+	public void injectServices(ServiceRegistryImplementor serviceRegistry) {
+		this.serviceRegistry = serviceRegistry;
+		if ( connectionProvider instanceof ServiceRegistryAwareService ) {
+			( (ServiceRegistryAwareService) connectionProvider ).injectServices( serviceRegistry );
+		}
+	}
+
+	@Override
+	public void configure(Map configurationValues) {
+		if ( connectionProvider instanceof Configurable ) {
+			Configurable configurableConnectionProvider = (Configurable) connectionProvider;
+			configurableConnectionProvider.configure( configurationValues );
+		}
+	}
+
+	@Override
+	public Connection getConnection() throws SQLException {
+		Connection connection;
+		try {
+			connection = connectionProvider.getConnection();
+		}
+		catch (Exception e) {
+			throw e;
+		}
+
+		// protect operation execution behind a double checked lock
+		// we don't want them to be applied multiple times
+		if ( operations != null ) {
+			// if lock is already held, we are reentering and acquiring a connection for the an operation execution
+			// so we need to ignore operations here
```

Comment:
Is there a strong guarantee that all "reentrant" operations are executed by the same thread?
Maybe a different kind of switch would make this code future-proof to concurrent initialization.